---
#Άσκηση 1. Επιβλεπόμενη Μάθηση: Ταξινόμηση
---

Συνεργάτες:
Δούλης Κωνσταντίνος 03116175

Καλογερόπουλος Ιωάννης 03116117

Κατσίκας Μουρούτσος Γεώργιος 03116132

---

Mικρό Dataset

Αρχικά, ενημερώνουμε τις βιβλιοθήκες που θα χρησιμοποιηθούν (έχει προστεθεί στην αρχή του κελιού η magic command %%capture ώστε να μην εμφανίζονται στο stdouput οι πληροφορίες των εγκαταστάσεων, για να είναι πιο ευανάγνωστο):

In [1]:
!pip install --upgrade pip #upgrade pip package installer
!pip install scikit-learn --upgrade #upgrade scikit-learn package
!pip install numpy --upgrade #upgrade numpy package
!pip install pandas --upgrade #--upgrade #upgrade pandas package
!pip install -U tensorflow

Requirement already up-to-date: pip in /usr/local/lib/python3.6/dist-packages (20.2.4)
Requirement already up-to-date: scikit-learn in /usr/local/lib/python3.6/dist-packages (0.23.2)
  Using cached numpy-1.19.4-cp36-cp36m-manylinux2010_x86_64.whl (14.5 MB)
  Attempting uninstall: numpy
    Found existing installation: numpy 1.18.5
    Uninstalling numpy-1.18.5:
      Successfully uninstalled numpy-1.18.5
ERROR: After October 2020 you may experience errors when installing or updating packages. This is because pip will change the way that it resolves dependency conflicts.

We recommend you use --use-feature=2020-resolver to test your packages with the new resolver before it becomes the default.

tensorflow 2.3.1 requires numpy<1.19.0,>=1.16.0, but you'll have numpy 1.19.4 which is incompatible.
datascience 0.10.6 requires folium==0.2.1, but you'll have folium 0.8.3 which is incompatible.
albumentations 0.1.12 requires imgaug<0.2.7,>=0.2.5, but you'll have imgaug 0.2.9 which is incompatib

Requirement already up-to-date: pandas in /usr/local/lib/python3.6/dist-packages (1.1.4)
Requirement already up-to-date: tensorflow in /usr/local/lib/python3.6/dist-packages (2.3.1)
  Using cached numpy-1.18.5-cp36-cp36m-manylinux1_x86_64.whl (20.1 MB)
  Attempting uninstall: numpy
    Found existing installation: numpy 1.19.4
    Uninstalling numpy-1.19.4:
      Successfully uninstalled numpy-1.19.4
ERROR: After October 2020 you may experience errors when installing or updating packages. This is because pip will change the way that it resolves dependency conflicts.

We recommend you use --use-feature=2020-resolver to test your packages with the new resolver before it becomes the default.

datascience 0.10.6 requires folium==0.2.1, but you'll have folium 0.8.3 which is incompatible.
albumentations 0.1.12 requires imgaug<0.2.7,>=0.2.5, but you'll have imgaug 0.2.9 which is incompatible.


In [2]:
import warnings 
warnings.filterwarnings('ignore')

In [3]:
from urllib.request import urlretrieve
import os
import zipfile

def download(url, file):
    if not os.path.isfile(file):
        urlretrieve(url,file)
        print("File downloaded")

download('http://archive.ics.uci.edu/ml/machine-learning-databases/credit-screening/crx.data', 'crx.data')

#Japanese Credit Screening


##Πληροφορίες dataset

Το dataset που εξετάζουμε παρέχει πληροφορίες σχετικά με άτομα στα οποία χορηγήθηκε ή όχι πίστωση από Ιαπωνική εταιρεία. Τα δεδομένα παράχθηκαν ύστερα από σχετικές ερωτήσεις στους πελάτες τις εταιρείας.

##Περιγραφή χαρακτηριστικών του dataset

To dataset διαθέτει <b>690 δείγματα</b>, με το καθένα από αυτά να διαθέτει <b>15 χαρακτηριστικά συν 1 που είναι η κλάση</b>.

Το κάθε δείγμα αντιστοιχεί σε μία γραμμή του αρχείου, ενώ η κάθε στήλη σε ένα χαρακτηριστικό του. <b>Η τελευταία στήλη δηλώνει την κλάση που ανήκει το δείγμα</b>. Συγκεκριμένα, το σύμβολο <b>"+"</b> δηλώνει ότι η πίστωση <b>εγκρίθηκε</b>, ενώ το σύμβολο <b>"-"</b> δηλώνει ότι <b>δεν εγκρίθηκε</b>.
Επομένως, η δομή του αρχείου είναι κατάλληλη για να ξεκινήσουμε την προεπεξεργασία του dataset.

Για λόγους απορήτου τα attribute names και values έχουν αντικατασταθεί με τυχαία σύμβολα, που δεν παρέχουν κάποια σχετική πληροφορία σχετικά με το τι περιγράφουν.

Το είδος των χαρακτηριστικών φαίνεται παρακάτω:

| Attribute | Type of Data                                 | Description    |
|-----------|----------------------------------------------|----------------|
| A1        | b, a.                                        | Male           |
| A2        | continuous.                                  | Age            |
| A3        | continuous.                                  | Debt           |
| A4        | u, y, l, t.                                  | Married        |
| A5        | g, p, gg.                                    | BankCustomer   |
| A6        | c, d, cc, i, j, k, m, r, q, w, x, e, aa, ff. | EducationLevel |
| A7        | v, h, bb, j, n, z, dd, ff, o.                | Ethnicity      |
| A8        | continuous.                                  | YearsEmployed  |
| A9        | t, f.                                        | PriorDefault   |
| A10       | t, f.                                        | Employed       |
| A11       | continuous.                                  | CreditScore    |
| A12       | t, f.                                        | DriversLicense |
| A13       | g, p, s.                                     | Citizen        |
| A14       | continuous.                                  | ZipCode        |
| A15       | continuous.                                  | Income         |
| A16       | +,-                                          | Approved (CLASS ATTRIBUTE)       |



* Τα χαρακτηριστικά που έχουν continuous τιμές αποτελουν <b>διατεταγμένα </b>χαρακτηριστικά, ενώ αυτά που έχουν διακριτές τιμές, δηλαδή σύμβολα, <b>μη διατεταγμένα</b>.

* Στο dataset <b>δεν</b> υπάρχουν επικεφαλίδες, καθώς ούτε και αρίθμηση γραμμών.

* Οι ετικέτες και οι σημασίες τους είναι:
    * "+": Εγκρίθηκε η πίστωση
    * "-": Απορρίφθηκε η πίστωση

    Οι ετικέτες αυτές βρίσκονται στην τελευταία στήλη, ενώ οι τιμές τους θα πρέπει να μετατραπούν όπως φαίνεται παρακάτω:
    * "+" -> 1
    * "-" -> 0

* <b>Υπάρχουν απουσιάζουσες τιμές</b>. Συγκεκριμένα, 37 δείγματα (5%) του συνολικού αριθμού δειγμάτων παρουσιάζουν μία ή παραπάνω απώλειες δεδομένων. 

Όλα τα παραπάνω φαίνονται παρακάτω:

##Προεπεξεργασία των δεδομένων
Aρχικά διαβάζουμε το αρχείο <b>crx.data</b>.

In [4]:
import pandas as pd

df = pd.read_csv("crx.data", header=None)
# print the five first samples
print(df[:5])

  0      1      2  3  4  5  6     7  8  9   10 11 12     13   14 15
0  b  30.83  0.000  u  g  w  v  1.25  t  t   1  f  g  00202    0  +
1  a  58.67  4.460  u  g  q  h  3.04  t  t   6  f  g  00043  560  +
2  a  24.50  0.500  u  g  q  h  1.50  t  f   0  f  g  00280  824  +
3  b  27.83  1.540  u  g  w  v  3.75  t  t   5  t  g  00100    3  +
4  b  20.17  5.625  u  g  w  v  1.71  t  f   0  f  s  00120    0  +


###Αντιστοίχιση κλάσεων
Παρατηρούμε, όπως ειπώθηκε και παραπάνω, ότι οι κλάσεις δηλώνονται στην τελευταία στήλη με τα σύμβολα +/-. 

Μετατρέπουμε τα "+" σε 1 και τα "-" σε 0:

In [5]:
# create mapper for each symbol
class_mapper = {"+": 1, "-": 0}
# since these symbols occur only on the last column we dont specify the column
df = df.replace(class_mapper)
print(df[:5])

  0      1      2  3  4  5  6     7  8  9   10 11 12     13   14  15
0  b  30.83  0.000  u  g  w  v  1.25  t  t   1  f  g  00202    0   1
1  a  58.67  4.460  u  g  q  h  3.04  t  t   6  f  g  00043  560   1
2  a  24.50  0.500  u  g  q  h  1.50  t  f   0  f  g  00280  824   1
3  b  27.83  1.540  u  g  w  v  3.75  t  t   5  t  g  00100    3   1
4  b  20.17  5.625  u  g  w  v  1.71  t  f   0  f  s  00120    0   1


###Χειρισμός απουσιάζουσων τιμών χαρακτηριστικών

Στη συνέχεια, θα εντοπίσουμε τις απουσιάζουσες τιμές. Αυτές δηλώνονται στο dataset με τον χαρακτήρα "?".

In [6]:
import numpy as np
from sklearn.impute import SimpleImputer

# replace "?" with np.NaN
df.replace('?',np.NaN,inplace=True)

# calculate the number of samples with at least one missing attribute
# df.isna() is the mask of df where each element is True if is NaN
# s is the length of the list of samples that have at least one True(NaN)
num_of_incomplete_samples = len([i for i in np.array(df.isna()) if True in i])
print("The samples of the dataset that have at least one missing attribute are ", num_of_incomplete_samples)
print("Which means that the ", num_of_incomplete_samples*100/df.shape[0], "% of the samples have missing values.", sep="")

The samples of the dataset that have at least one missing attribute are  37
Which means that the 5.36231884057971% of the samples have missing values.


Από τα παραπάνω βλέπουμε ότι αν θέλαμε να αγνοήσουμε τα δείγματα που έχουν missing values θα έπρεπε να "πετάξουμε" το 5% των δειγμάτων, αριθμός που <b>θα επηρέαζε αρνητικά τα αποτελέσματα</b>.

Προκειμένου να εξετάσουμε αν οι περισσότερες τιμές που λείπουν προέρχονται απο συγκεκριμένα (λίγα) χαρακτηριστικά (και αν αυτά έχουν ίσως και μηδενική διακύμανση, δηλαδή  χαρακτηριστικά με σταθερές τιμές), οπότε θα μπορούσαμε να τα αγνοήσουμε, εκτελούμε:

In [7]:
print(df.shape)

(690, 16)


In [8]:
# the list of the number of missing values for each attribute occurs by 
# summing the elements of the inverse of df, without df's last column.
# df's last column is the class attribute and it is alwasy present.
incomplete_attrs = [sum(i) for i in np.array(df.isna())[:,:df.shape[1]-1].T]
print("For each attribute of the dataset, the number of the missing values is")
print(incomplete_attrs)
# print(sum(incomplete_attrs))

For each attribute of the dataset, the number of the missing values is
[12, 12, 0, 6, 6, 9, 9, 0, 0, 0, 0, 0, 0, 13, 0]


Βλέπουμε, επομένως, ότι από τα 690 δείγματα το πολύ σε 13 από αυτά ένα χαρακτηριστικό δεν έχει τιμές. Επομένως, δεν έχει νόημα να αγνοήσουμε το χαρακτηριστικό αυτό, ακόμα και αν είναι αυτό με τις περισσότερες ελλείψεις, αφού αυτές είναι πολύ λίγες και θα χάναμε πολύ πληροφορία.

Τελικά η πιο συμφέρουσα λύση είναι να αντικαταστήσουμε τα NaN με τιμές που προκύπτουν με βάση τις τιμές των χαρακτηριστικών σε προηγούμενα δείγματα. <u>Συγκεκριμένα, αντικαθιστούμε τα NaN με την τιμή του χαρακτηριστικού που εμφανίζεται πιο συχνά</u>.

In [9]:
# create imputer that will replace Nan with the most frequent value
imp=SimpleImputer(missing_values=np.NaN,strategy="most_frequent")
idf=pd.DataFrame(imp.fit_transform(df))
idf.columns=df.columns
idf.index=df.index
print(idf[:5])

  0      1      2  3  4  5  6     7  8  9  10 11 12     13   14 15
0  b  30.83      0  u  g  w  v  1.25  t  t  1  f  g  00202    0  1
1  a  58.67   4.46  u  g  q  h  3.04  t  t  6  f  g  00043  560  1
2  a  24.50    0.5  u  g  q  h   1.5  t  f  0  f  g  00280  824  1
3  b  27.83   1.54  u  g  w  v  3.75  t  t  5  t  g  00100    3  1
4  b  20.17  5.625  u  g  w  v  1.71  t  f  0  f  s  00120    0  1


Οι κλάσεις βρίσκονται στην τελευταία στήλη του συνόλου δεδομένων. Οπότε για να βρούμε το πλήθος τους και τα ποσοστά δειγμάτων τους επί του συνόλου:

In [10]:
num_of_rows = idf.shape[0]
num_of_attrs = idf.shape[1] - 1 #remove one element because of the class attribute

# get labesl and features
labels_df = idf.iloc[:, [num_of_attrs]] # τα labels είναι στην τελευταία κολώνα
features_df = idf.iloc[:, 0:num_of_attrs]  # τα features είναι όλες οι προηγούμενες κολώνες

labels = labels_df.values.reshape(num_of_rows,)
features = features_df.values
# convert to int
labels.astype(int)
labels = np.array(labels, dtype='int64')
# print(labels.shape)

# find how many of each class
bin_count = np.bincount(labels)
print ("frequencies:", bin_count)
print("The percentage of 0's in data: ", bin_count[0]*100/sum(bin_count), "%.")
print("The percentage of 1's in data: ", bin_count[1]*100/sum(bin_count), "%.")

frequencies: [383 307]
The percentage of 0's in data:  55.507246376811594 %.
The percentage of 1's in data:  44.492753623188406 %.


Καταλήγουμε, λοιπόν, ότι τα ποσοστά είναι περίπου 55.5% για την κλάση 0 και 44.5% για την κλάση 1. Επομένως, το dataset είναι <b>ισορροπημένο</b>.

In [11]:
# print(labels[:5])
print(features.shape)

(690, 15)


###Χειρισμός κατηγορικών χαρακτηριστικών
Το τελευταίο βήμα πριν τον διαχωρισμό του dataset είναι να μετατρέψουμε τα μη διατεταγμένα χαρακτηριστικά  με m τιμές σε m binary χαρακτηριστικά από τα οποία μόνο ένα είναι ενεργό κάθε φορά:

In [12]:
# μετατρέπουμε σε dataframe και τυπώνουμε
mtdf = pd.DataFrame(features)

# οι κολόνες 1 και 4 έχουν κατηγορικές μεταβλητές. 
# Με την "get_dummies" κάνουμε τη μετατροπή σε binary χαρακτηριστικά που περιγράψαμε
dummies = pd.get_dummies(mtdf, columns=[0,3,4,5,6,8,9,11,12])
# print(dummies)

# Μετατρέπουμε σε αριθμητικές τιμές (pd.to_numeric) και σε numpy array (.values)
all_data = dummies.apply(pd.to_numeric).values
print(all_data.shape)
# print(all_data)

(690, 46)


Βλέπουμε, όπως αναμενόταν, ότι το πλήθος των χαρακτηριστικών φαίνεται έχει πλέον αυξηθεί.
Στην πραγματικότητα παραμένουν τα ίδια χαρακτηριστικά, ωστόσο για κάθε κατηγορικό χαρακτηριστικό Κ, δημιουργούμε το χαρακτηριστικό έχειΚ_V για κάθε δυνατή τιμή του K, V, το οποίο έχει τιμή 1 μόνο αν το συγκεκριμένο δείγμα έχει την τιμή V για το χαρακτηριστικό αυτό.

##Διαχωρισμός του dataset
Διαχωρίζουμε το dataset σε train και test set, χρησιμοποιώντας το 80% των δεδομένων για το training και το 20% για το testing:

In [13]:
from sklearn.model_selection import train_test_split


train, test, train_labels, test_labels = train_test_split(all_data, labels, test_size=0.2, random_state=78)

In [14]:
from collections import defaultdict
# just for checking...
print(train.shape)
print(test.shape)
check = defaultdict(int)
for i in range(0,len(train_labels)):
    check['pos_train'] += train_labels[i] == 1
    check['neg_train'] += train_labels[i] == 0
    

# for i in range(0,len(test_labels)):
#     check['pos_test'] += test_labels[i] == 1
#     check['neg_test'] += test_labels[i] == 0
    
# print(check)

(552, 46)
(138, 46)


##Ορισμός μετασχηματιστών

###Variance Threshold

Παρακάτω υλοποιείται η συνάρτηση που αρχικοποιεί selector Variance Threshold, με κατόφλι που δίνεται από το όρισμα threshold

In [15]:
from sklearn.feature_selection import VarianceThreshold

def Variance_Threshold(data, threshold=0):
    # αρχικοποιούμε έναν selector
    train, test, train_labels, test_labels = data
    selector = VarianceThreshold(threshold = threshold)
    train_reduced = selector.fit_transform(train)
    test_reduced = selector.transform(test)

    return (train_reduced, test_reduced, train_labels, test_labels)

###PCA

In [16]:
from sklearn.decomposition import PCA

def PCA_selector(data, n):
    """
    Function that creates a PCA selector that learns the train data
    and transforms train and test data.
    """
    train, test, train_labels, test_labels = data
    # αρχικοποιούμε τον selector
    pca = PCA(n_components=n)
    trainPCA = pca.fit_transform(train)
    testPCA = pca.transform(test)

    return (trainPCA, testPCA, train_labels, test_labels)

###Κανονικοποίηση

In [17]:
from scipy import stats as st

def StandarizerTransformer(data, method):
    """
    Function that standarizes the data using two different methods.
    minmax and zscore.    
    """
    train, test, train_labels, test_labels = data
    if method=="zscore":
        std_train = st.zscore(train)
    elif method=="minmax":
        std_train = (train - np.min(train) )/ (np.max(train) - np.min(train))
    else:
        std_train = train
    return (std_train, test, train_labels, test_labels)

###Εξισορρόπηση dataset - Sampling

In [18]:
!pip install six

In [19]:
# from sklearn.externals import six
# from imblearn.over_sampling import RandomOverSampler
# from imblearn.under_sampling import RandomUnderSampler

def sampling(data, method):
    train, test, train_labels, test_labels = data
    if method == "over":
        ros = RandomOverSampler()
        # mlb = MultiLabelBinarizer().fit(['0', '1', '2', '3', '4'])
        # tmplabels = np.asarray(mlb.inverse_transform(C_trainDataTargets)).flatten()
        train_resampled, train_labels_resampled = ros.fit_sample(train,train_labels)
        trainTargets_resampled = mlb.transform(trainTargets_resampled)
    elif method == 'under':
        ros = RandomUnderSampler()
        # mlb = MultiLabelBinarizer().fit(['0', '1', '2', '3', '4'])
        # tmplabels = np.asarray(mlb.inverse_transform(C_trainDataTargets)).flatten()
        train_resampled, train_labels_resampled = ros.fit_sample(train,train_labels)
        trainTargets_resampled = mlb.transform(trainTargets_resampled)
    
    return (train_resampled, test, train_labels_resampled, test_labels)

##Ορισμός ταξινομητών

In [20]:
from sklearn.dummy import DummyClassifier
from sklearn.naive_bayes import GaussianNB


def createClassifier(classifier, argument=5):

    if classifier == 'dummy':
        clf = DummyClassifier(strategy=argument)
    elif classifier == 'gnb':
        clf = GaussianNB()
    elif classifier == 'knn':
        clf = KNeighborsClassifier(n_neighbors=argument)
    
    return clf

##Ορισμός Pipeline

In [21]:
def Pipeline (data, steps):
    """
    Args:
        * data: (train, train_labels, test, test_labels)
        * steps: list of tuples, where each tuple's 1st el 
        is the transformer/classifier and the 2nd el
        is the arguments of the function for the
        transformer/classifier
    """

    mapper = {
        'selector': Variance_Threshold,
        'scaler': StandarizerTransformer,
        'sampler': sampling,
        'pca': PCA_selector,
        'kNN': createClassifier,
        'dummy': createClassifier
    }

    classifiers = set(['kNN', 'uniform', 'constant_0', 'constant_1', 'most_frequent', 'stratified'])

    for step in steps:
        if step[0] not in classifiers:
            # update data
            data = mapper[step[0]](step[1])

        else:
            # return model
            return (data, createClassifier(step[0], step[1]))




In [22]:
!pip install --upgrade imbalanced-learn

     |████████████████████████████████| 167 kB 5.8 MB/s 
  Attempting uninstall: imbalanced-learn
    Found existing installation: imbalanced-learn 0.4.3
    Uninstalling imbalanced-learn-0.4.3:
      Successfully uninstalled imbalanced-learn-0.4.3


In [31]:
train_variance = train.var(axis=0)
print(train_variance)
print(np.max(train_variance))
print(train_variance.shape)

[1.37390925e+02 2.50808996e+01 1.13895028e+01 2.54195514e+01
 3.14871660e+04 1.31891623e+07 2.09564036e-01 2.09564036e-01
 1.80831233e-03 1.76157845e-01 1.75170001e-01 1.76157845e-01
 1.80831233e-03 1.75170001e-01 7.18303665e-02 1.69105099e-01
 5.46103760e-02 3.99305556e-02 3.65961720e-02 6.56604442e-02
 8.08883375e-02 1.60385161e-02 6.72127704e-02 5.62086484e-02
 9.54598561e-02 3.61006091e-03 8.23750263e-02 5.62086484e-02
 8.08883375e-02 1.07514178e-02 7.33564377e-02 1.61730729e-01
 1.25203476e-02 5.40524575e-03 1.80831233e-03 2.42438563e-01
 1.07514178e-02 2.49051539e-01 2.49051539e-01 2.45746692e-01
 2.45746692e-01 2.47948829e-01 2.47948829e-01 8.53287125e-02
 8.97592418e-03 7.78952689e-02]
13189162.332936361
(46,)


In [32]:
for sample in train:
    print(sample[5])

0.0
9.0
0.0
500.0
1000.0
0.0
6.0
1.0
18.0
0.0
0.0
0.0
2079.0
15.0
108.0
0.0
15.0
0.0
126.0
19.0
7.0
2.0
1210.0
27.0
0.0
0.0
38.0
20.0
0.0
0.0
173.0
0.0
316.0
2197.0
3000.0
0.0
21.0
0.0
18.0
2000.0
0.0
9800.0
0.0
1.0
500.0
0.0
0.0
196.0
0.0
0.0
0.0
0.0
50.0
0.0
456.0
0.0
1200.0
3.0
500.0
67.0
475.0
0.0
16.0
283.0
300.0
0.0
98.0
0.0
0.0
0.0
1000.0
500.0
0.0
1210.0
0.0
1000.0
0.0
400.0
540.0
0.0
0.0
0.0
50.0
28.0
1.0
0.0
3.0
540.0
0.0
200.0
108.0
610.0
1.0
300.0
0.0
1.0
0.0
1430.0
0.0
1187.0
1300.0
0.0
200.0
809.0
0.0
251.0
321.0
2.0
0.0
0.0
0.0
23.0
2.0
0.0
0.0
100.0
2100.0
109.0
0.0
40.0
600.0
237.0
300.0
6.0
0.0
0.0
0.0
2279.0
0.0
120.0
20.0
0.0
0.0
0.0
1.0
350.0
0.0
50000.0
154.0
150.0
0.0
0.0
0.0
0.0
0.0
0.0
1260.0
4.0
300.0
687.0
0.0
0.0
0.0
0.0
1249.0
50.0
0.0
0.0
0.0
51100.0
5.0
17.0
1000.0
0.0
3000.0
0.0
0.0
0.0
0.0
246.0
228.0
1.0
99.0
4.0
14.0
800.0
560.0
13.0
0.0
247.0
12.0
948.0
1.0
204.0
4.0
22.0
32.0
0.0
824.0
0.0
0.0
1391.0
100.0
0.0
6.0
0.0
0.0
0.0
0.0
5860.0
0.0
0.0
0.0


In [35]:
t = [1,2,3,4,5]
for w in range(0,len(t)):
    for i in range(0,w):
        print(t[i:w])

[1]
[1, 2]
[2]
[1, 2, 3]
[2, 3]
[3]
[1, 2, 3, 4]
[2, 3, 4]
[3, 4]
[4]


In [38]:
sel = [True, False]
scal = [True, False]
_pca = [True, False]
for si in sel:
    for scali in scal:
        for pcai in _pca:
            if si: print("si", end="\t")
            if scali: print("scali", end="\t")
            if pcai: print("pcai")

si	scali	pcai
si	scali	si	pcai
si	scali	pcai
scali	pcai


In [78]:
from itertools import combinations

def get_transformations(my_list):
    sublist = []
    for i in range(0, len(my_list)+1):
        temp = [list(x) for x in combinations(my_list, i)]  #get the combinations with i elements
        if not (('scaler' in temp and 'min_max_scaler') or ('ros' in temp and 'rus' in temp)):
            sublist.extend(temp)    #and add to initial list
    return sublist



trans = ['selector', 'scaler', 'sampler', 'pca']
print(get_transformations(trans))

[[], ['selector'], ['scaler'], ['sampler'], ['pca'], ['selector', 'scaler'], ['selector', 'sampler'], ['selector', 'pca'], ['scaler', 'sampler'], ['scaler', 'pca'], ['sampler', 'pca'], ['selector', 'scaler', 'sampler'], ['selector', 'scaler', 'pca'], ['selector', 'sampler', 'pca'], ['scaler', 'sampler', 'pca'], ['selector', 'scaler', 'sampler', 'pca']]


In [114]:
a = {1:11, 2:22}
if 1 in a: print(a.keys())


dict_keys([1, 2])


In [126]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.utils import shuffle
from sklearn import neighbors
from sklearn.metrics import classification_report

from imblearn.pipeline import Pipeline

from sklearn.dummy import DummyClassifier
from sklearn.feature_selection import VarianceThreshold
from sklearn.preprocessing import StandardScaler # φέρνουμε τον StandarScaler ως transformer που έχει .transform kai ΄όχι ως scale()
from imblearn.over_sampling import RandomOverSampler
from imblearn.under_sampling import RandomUnderSampler
from sklearn.decomposition import PCA
from sklearn.model_selection import GridSearchCV
from sklearn import preprocessing
from sklearn.naive_bayes import GaussianNB



selector = VarianceThreshold()
min_max_scaler = preprocessing.MinMaxScaler()
scaler = StandardScaler()
ros = RandomOverSampler()
rus = RandomUnderSampler()
pca = PCA()
dummy = DummyClassifier()
gnb = GaussianNB()
knn = neighbors.KNeighborsClassifier(n_jobs=-1) # η παράμετρος n_jobs = 1 χρησιμοποιεί όλους τους πυρήνες του υπολογιστή


vthresholds = [0, 1000, 10000, 1000000] #προσαρμόζουμε τις τιμές μας στο variance που παρατηρήσαμε
scalers = [None, min_max_scaler, scaler]
samplers = [None, ros, rus]
num_n_components = [10, 20, 30, 40, 50, 60]

k = [1, 6] # η υπερπαράμετρος του ταξινομητή
scoring = ['f1_macro', 'f1_micro']

transformers = {'selector': selector,
         'scaler': scaler,
         'min_max_scaler': min_max_scaler,
         'ros': ros,
         'rus': rus,
         'pca': pca
        }

classifiers = {
    'dummy': dummy,
    'gnb': gnb,
    'knn': knn

}

est_mapper_keys = {
    'selector': 'selector__threshold',
    'pca': 'pca__n_components',
    'dummy': 'dummy__strategy',
    'knn': 'kNN__n_neighbors'
}


est_values_mapper = {
    'selector': [0, 1000, 10000, 1000000],
    'pca': [10, 20, 30, 40, 50, 60],
    'dummy': ['uniform', 'constant_0', 'constant_1', 'most_frequent', 'stratified'],
    'knn': [1, 6]
}

def getEstDict():
    dict = {}
    for step in steps:
        if step[0] in est_mapper_keys:
            dict[est_mapper_keys[step[0]]] = est_values_mapper[step[0]]
    return dict


for sequence in get_transformations(transformers.keys()):
    # add trasformers
    steps = [(trans, transformers[trans]) for trans in sequence]
    # add classifier
    for classifier in classifiers.keys():
        if len(steps): steps.pop()  #delete previous classifier
        steps.append((classifier, classifiers[classifier])) #add new classifier
        # create Pipe
        pipe = Pipeline(steps=steps, memory = 'tmp')
        estimator_dict = getEstDict()
        # print(estimator_dict)
        # break
        estimator = GridSearchCV(pipe, dict(selector__threshold=vthreshold, pca__n_components=n_components, dummy__strategy=['uniform', 'constant_0', 'constant_1', 'most_frequent', 'stratified']), cv=10, scoring='f1_macro', n_jobs=-1)
    # break


# import time
# start_time = time.time()
# estimator.fit(train, train_labels)
# preds = estimator.predict(test)
# print("Συνολικός χρόνος fit και predict: %s seconds" % (time.time() - start_time))
# print(classification_report(test_labels, preds))

In [24]:
print(estimator.best_estimator_)
print(estimator.best_params_)

Pipeline(memory='tmp',
         steps=[('selector', VarianceThreshold(threshold=0)),
                ('scaler', StandardScaler()), ('sampler', RandomOverSampler()),
                ('pca', PCA(n_components=10)),
                ('dummy', DummyClassifier(strategy='stratified'))])
{'dummy__strategy': 'stratified', 'pca__n_components': 10, 'selector__threshold': 0}
